File: 2015/python/1/main.py
#!/usr/bin/env python3.12.3
import os
import sys

if __name__ == '__main__':
    args = sys.argv

    if len(args) < 2:
        print('Usage: python main.py <file>')
        sys.exit(1)

    file = args[1]

    if not os.path.exists(file):
        print('File not found')
        sys.exit(1)

    file_content = None
    with open(file) as f:
        file_content = f.read()

    floor = 0
    count = 0
    basement = 0
    for i, c in enumerate(file_content):
        if c == '(':
            floor += 1
        elif c == ')':
            floor -= 1

        count += 1
        if floor == -1 and basement == 0:
            basement = count

    print("Part 1:", floor)
    print("Part 2:", basement)

File: 2015/ruby/1/main.rb
#!/usr/bin/env ruby

def main()
    args = ARGV
    if args.length < 1
        puts "Usage: ruby main.rb <input.txt>"
        exit 1
    end

    filename = args[0]
    file = File.open(filename, "r")

    if file.nil?
        puts "Could not open file"
        exit 1
    end

    lines = file.readlines
    file.close

    floor = 0
    basement = 0
    count = 0
    lines.each do |line|
        line.each_char do |char|
            count += 1
            if char == '('
                floor += 1
            elsif char == ')'
                floor -= 1
            end

            if floor == -1 && basement == 0
                basement = count
            end
        end
    end

    puts "Part 1: #{floor}"
    puts "Part 2: #{basement}"
end

main()